trax/models/research/funnel_transformer.py

```diff
@@ -0,0 +1,269 @@
+                             dropout=0.1,
+                             dropout_shared_axes=None,
+                             mode='train',
+                             ff_activation=tl.Relu,
+                             pool_layer=tl.AvgPool,
+                             pool_size=(2,),
+                             strides=(2,),
+                             separate_cls=True):
+  """Returns a Funnel Encoder.
+  """
+  assert encoder_segment_lengths
+
+  positional_encoder = [
+      tl.Embedding(vocab_size, d_model),
+      tl.Dropout(rate=dropout, shared_axes=dropout_shared_axes, mode=mode),
+      tl.PositionalEncoding(max_len=max_len)]
+
+  encoder_blocks = []
+  n_encoder_segments = len(encoder_segment_lengths)
+
+  for i in range(n_encoder_segments):
+    # Building i'th segment
+    for _ in range(encoder_segment_lengths[i]):
+      # Create segment_size encoder blocks
+      encoder_blocks.append(
+          _EncoderBlock(d_model, d_ff, n_heads, dropout, dropout_shared_axes,
+                        mode, ff_activation))
+
+    # If not last segment, add funnel block
+    if i != n_encoder_segments - 1:
+      encoder_blocks.append(_FunnelBlock(d_model, d_ff, n_heads, dropout,
+                                         dropout_shared_axes, mode,
+                                         ff_activation, pool_layer, pool_size,
+                                         strides, separate_cls))
+
+  # Assemble and return the model.
+  return tl.Serial(                               # toks
+      # Encode.
+      tl.Branch(
+          positional_encoder, tl.PaddingMask()),  # vecs masks
+      encoder_blocks,                             # vecs masks
+      tl.Select([0], n_in=2),                     # vecs
+      tl.LayerNorm(),                             # vecs
+
+      # Map to output categories.
+      tl.Mean(axis=1),                            # vecs
+      tl.Dense(n_classes),                        # vecs
+      tl.LogSoftmax(),                            # vecs
+  )
+
+
+def _FunnelResidualBlock(d_model, d_ff, n_heads,
+                         dropout, dropout_shared_axes, mode, ff_activation,
+                         pool_layer, pool_size, strides):
+  feed_forward = _FeedForwardBlock(
+      d_model, d_ff, dropout, dropout_shared_axes, mode, ff_activation)
+
+  dropout_ = tl.Dropout(
+      rate=dropout, shared_axes=dropout_shared_axes, mode=mode)
+
+  attn_ = tl.AttentionQKV(d_model, n_heads=n_heads, dropout=dropout, mode=mode)
+
+  pooling_ = PoolLayer(pool_layer, pool_size, strides)
+
+  return [
+      tl.Parallel(tl.Branch(pooling_, None), None),
+      tl.Residual(
+          tl.Parallel(tl.LayerNorm(), tl.LayerNorm()),
+          tl.Select([0, 1, 1, 2]),
+          attn_,
+          tl.Parallel(None, tl.Fn(
+              'mask_max_pool', _InternalMaxPool)),
+          dropout_
+      ),
+      tl.Residual(
+          feed_forward
+      )
+  ]
+
+
+def FunnelTransformer(vocab_size,
```

Comment:
Can we rename it to FunnelTransformerDecoder, to keep naming consistent with models/transformer.py ? It seems closer to TransformerDecoder than Transformer, since the former outputs an embedding per token (like this Funnel class) and the latter predicts a class per token.

Reply:
As per our previous discussion, we changed the `FunnelTransformer` to output token-level categorical distribution over vocab instead of embeddings, which makes it useful for example as a BERT.